### PR TITLE
attach drag behaviour to event rectangle

### LINKED ISSUE #114: attach drag behaviour to event rectangle

### PR DESCRIPTION
**Pull request by umarmohammed**
closes #114

### CODE REVIEW

File: projects/ngx-d3timeline/src/lib/content/content.service.ts [hunk 1]
```diff
@@ -125,4 +146,52 @@ export class ContentService {
...
+    return {
+      ...scales,
+      state: {
+        ...scales.state,
+        data
+      }
+    };
+  }
+
+  private shiftDraggedTimelineEvent(
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/content.service.ts]:
```suggestion
  private updateDraggedTimelineEvent(
```

File: projects/ngx-d3timeline/src/lib/content/content.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 export interface EventRectangle {
+  id: string | number;
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/content.ts:2]: what do you think about adding a type for this? It's already repeated here and in next file.

``` type identifier = string | number;```

File: projects/ngx-d3timeline/src/lib/content/drag.service.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    scan((acc, curr) => (curr ? { ...curr, dy: acc.dy + curr.dy } : acc), {
+      dx: 0,
+      dy: 0,
+      id: null
+    }),
+    takeUntil(this.dragEndSubject),
+    repeat()
+  );
+
+  onDrag(eventRectangle: EventRectangle, event: any) {
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/drag.service.ts]:
```suggestion
  onDrag(eventRectangle: EventRectangle, event: EventRectangleDragEvent) {
```

File: projects/ngx-d3timeline/src/lib/content/event-rectangle.component.ts [hunk 4]
```diff
@@ -25,6 +38,24 @@ import { EventRectangle } from './content';
...
+
+  constructor(private dragService: DragService) {}
+
+  ngAfterViewInit() {
+    this.setupDrag();
+  }
+
+  private setupDrag() {
+    if (this.eventRectangleEl) {
+      const onDrag = drag()
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/event-rectangle.component.ts:54]: can we use https://rxjs.dev/api/index/function/fromEventPattern in some way here?

File: projects/ngx-d3timeline/src/lib/content/content.component.spec.ts [hunk 5]
```diff
@@ -39,12 +39,14 @@ describe('ContentComponent', () => {
   it('should render correctly', () => {
     const rectangles: EventRectangle[] = [
       {
+        id: 1,
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/content.component.spec.ts:42]: doesn't this affect the snapshots?

File: projects/ngx-d3timeline/src/lib/content/drag.service.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+@Injectable({ providedIn: 'root' })
+export class DragService {
+  private dragSubject = new BehaviorSubject<EventRectangleDragEvent>(null);
+  private dragEndSubject = new Subject<any>();
+
+  drag$ = this.dragSubject.pipe(
+    scan(
+      (acc, curr) =>
+        curr ? { ...curr, dy: acc.dy + curr.dy, dx: acc.dx + curr.dx } : acc,
+      {
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/drag.service.ts]: could declutter this observable/scan by declaring the initial value in its own variable ie:

```
const dragSeed = {
  dx: 0,
  dy: 0,
  id: null
}
```

File: projects/ngx-d3timeline/src/lib/content/event-rectangle.component.spec.ts [hunk 7]
```diff
@@ -22,6 +22,7 @@ describe('EventRectangleComponent', () => {
 
   it('should render correctly', () => {
     fixture.componentInstance.eventRectangle = {
+      id: 1,
```
Comment (wflemingnz-retired) [inline projects/ngx-d3timeline/src/lib/content/event-rectangle.component.spec.ts:25]: does this change the snapshot?

**Review by wflemingnz-retired — CHANGES_REQUESTED** [review]

File: projects/ngx-d3timeline/src/lib/content/event-rectangle.component.spec.ts (hunk 7 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/event-rectangle.component.spec.ts:25]: No the id doesn't get rendered in the html

File: projects/ngx-d3timeline/src/lib/content/content.component.spec.ts (hunk 5 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/content.component.spec.ts:42]: No the id doesn't get rendered in the html

File: projects/ngx-d3timeline/src/lib/content/content.service.ts (hunk 1 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/content.service.ts]: done

File: projects/ngx-d3timeline/src/lib/content/content.ts (hunk 2 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/content.ts:2]: It also needs to be in TimelineEvent, might break a few things. I've created a new issue for this #117

File: projects/ngx-d3timeline/src/lib/content/drag.service.ts (hunk 6 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/drag.service.ts]: done

File: projects/ngx-d3timeline/src/lib/content/drag.service.ts (hunk 3 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/drag.service.ts]: done

File: projects/ngx-d3timeline/src/lib/content/event-rectangle.component.ts (hunk 4 above)
Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/event-rectangle.component.ts:54]: Good find, although I'm not sure that it will simplify the code.

Comment (umarmohammed) [reply projects/ngx-d3timeline/src/lib/content/event-rectangle.component.ts:54]: I've created an issue to investigate #118

Comment (wflemingnz-retired) [reply projects/ngx-d3timeline/src/lib/content/event-rectangle.component.ts:54]: yeah not sure either, could be worth keeping it in mind. I used it for a Titan related POC to hook into the map events (I think).

**Review by wflemingnz-retired — APPROVED** [review]

**Pull request merged**